### ISSUE TITLE
Not all arguments converted during string formatting - Hassio ERROR

### ISSUE THREAD

**Issue by vincenzosuraci** [peterbuga/HASS-sonoff-ewelink#89]
I'm not sure if it is the cause, however, looking at the log, once I get this message:

    2019-04-14 12:08:11 ERROR (Thread-2) [custom_components.sonoff] websocket error: [Errno 104] Connection reset by peer

Hassio hangs and the lovelace UI stops working, while other services (e.g. SSH, Samba, etc.) continue to operate.

**Comment by peterbuga** [comment]
do you have the latest code from `master`? or are you on older polling version? if it's the latter i strongly advice you to increase the `scan_interval` value or update the component to the latest version

**Comment by vincenzosuraci** [comment]
Thanks for your prompt reply. 
I got the message: Sonoff custom-component files updated.

**Comment by vincenzosuraci** [comment]
Did it. Now, no hangs but a lot of these ERROR (approx, one each 2 seconds):

    2019-04-14 19:47:11 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x6fd50f50>>: not all arguments converted during string formatting

**Comment by peterbuga** [comment]
i might have a faint idea why this is happening, but in the same time it might be some weird configuration on your side (because the code has been reported to work just fine right now) i really have no idea right now. 

unfortunately being a custom component i'm not actively supporting HASSIO installs nor i will anytime soon. depending on the devices you have, you could also give a try to this [release](https://github.com/peterbuga/HASS-sonoff-ewelink/releases/tag/v0.3.2)

ps: it wouldn't hurt also if you could provide the devices output as explained [here](https://github.com/peterbuga/HASS-sonoff-ewelink/tree/master/sonoff-debug) or follow the steps related to `debug` section in project's documentation.

**Comment by mattsaxon** [comment]
Peter, I'm getting the error too, though much less frequently.

`2019-04-15 19:14:21 ERROR (Thread-3) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x70203930>>: not all arguments converted during string formatting`

I had just updated to your latest version (this is must better performing, thank you for your hard work on this). Previously I was running your first commit and hadn't checked the error logs (I'm just getting my system up and running so am flushing through a number of issues with the wider config.

I've run the debug script you requested and here is the output.
[devices.json.txt](https://github.com/peterbuga/HASS-sonoff-ewelink/files/3081818/devices.json.txt)

The message appearing coincides with a 'hangs' in the HA UI where the switch icon takes a long time to respond.

Here is what I noted;

1. I operated switch in HA UI
2. Physical relay does not operate
3. Delay of approx 10 seconds
3. Switch in eWeLink app becomes unavailable (note, I'm using a different user id between HA and eWeLink so that is hopefully not the issue)
4. Delay of approx 10 seconds
5. Error message appears in HA log & Switch recovers in eWeLink (exactly the same time)

Usually after this, things are working again (though i sometime get back to back errors). I'm only operating a single switch here during this testing (despite the many in my config). They are all of model SOnOff S26 and I can reproduce with multiple of the switches (at least 2 so far).

**Comment by vincenzosuraci** [comment]
Some additional info gathered using HA instead of Hassio:

    2019-04-15 21:03:41 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x08C694F0>>: not all arguments converted during string formatting
  File "c:\users\vincenzo\homeassistant\lib\site-packages\websocket\_app.py", line 343, in _callback
    callback(*args)
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 311, in on_message
    self.set_entity_state(data['deviceid'], data['params']['switch'])
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 334, in set_entity_state
    '_'+str(outlet+1) if outlet is not None else ''

**Comment by mattsaxon** [comment]
I've an even simpler recreation now. If I use the eWeLink application to change the switch state, I get the failure 100% of the time.

If I close the eWeLink app, I get it much less regularly..... both eWeLink/SOnOff cloud and HA accessing the socket at the same time??

I get this error as well (i haven't managed to correlate it exactly as yet with what is triggering it)
`ERROR (Thread-3) [custom_components.sonoff] websocket error: Connection is already closed.`

**Comment by peterbuga** [comment]
guys i have to admit, these errors are a bit weird and somehow unrelated (obviously until i figure what goes wrong :P) 
i cannot do any updating right but i just noticed a small missed formatting (but then again, i'm not sure why it was working on the first place), can any of you test this little change:

on file `__init__.py` line 331 change this part
`entity_id = 'switch.%s%s' % (` to `entity_id = 'switch.%s%s%s' % (`
notice there's an extra **%s** 

and let me know how it goes! if this is the problem sorry for this, i must've been tired ¯\_(ツ)_/¯  i usually work on this in the nights after work

**Comment by vincenzosuraci** [comment]
@mattsaxon: I have a solution for your disconnection issue. 

It Is known that eWeLink server does not accept the same account to be connected simultaneously with different sockets to the server.

So I created a new eWeLink account with an e-mail alias (i use Gmail) and shared the devices with It. 

The drawback is that the first time you have to accept the invitation using the eWeLink app at the same time the invitation has been sent. 

I solved using and Android simulator in parallel with my smartphone. With the smartphone I invited the account I had on the simulator. So I have been able to activate the device sharing. On hassio I have the credentials of the alias. The, you can disinstall the simulator.

Result: I have no disconnections even if playing simultaneously (app + hassio).

**Comment by mattsaxon** [comment]
@vincenzosuraci, thanks for the suggestion, but unfortunately I already have separate accounts for eWeLink and HA

**Comment by mattsaxon** [comment]
@peterbuga that improves the situation, thanks, no red errors in the log. Note, its uncovered the next issue, which I will raise as a separate issue

**Comment by vincenzosuraci** [comment]
@peterbuga, many thanks for your suggestion!
I confirm that the suggested change did the trick. 

With the current version of the committed code I have:

```
2019-04-16 14:05:42 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x07D033F0>>: not all arguments converted during string formatting
  File "c:\users\vincenzo\homeassistant\lib\site-packages\websocket\_app.py", line 343, in _callback
    callback(*args)
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 311, in on_message
    self.set_entity_state(data['deviceid'], data['params']['switch'])
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 334, in set_entity_state
    '_'+str(outlet+1) if outlet is not None else ''
2019-04-16 14:07:42 ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x07D033F0>>: not all arguments converted during string formatting
  File "c:\users\vincenzo\homeassistant\lib\site-packages\websocket\_app.py", line 343, in _callback
    callback(*args)
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 311, in on_message
    self.set_entity_state(data['deviceid'], data['params']['switch'])
  File "C:\Users\Vincenzo\AppData\Roaming\.homeassistant\custom_components\sonoff\__init__.py", line 334, in set_entity_state
    '_'+str(outlet+1) if outlet is not None else ''
```
Instead, if I do the suggested change directly in the python code, the ERRORs disappear. 

I think that the source code line modification did the trick: indeed, you was passing 3 parameters, not 2.
I would suggest to commit the change, since it makes sense.

VS

**Comment by peterbuga** [comment]
guys thanks for reporting the problem and testing it! i pushed the fix. i'll close this issue for now but feel free to re-open it if the problem occurs.

**Comment by Tomelin** [comment]
HI @peterbuga,

I have a problem with my HA and Sonoff.  

How your resolved?

**Comment by Tomelin** [comment]
> HI @peterbuga,
> 
> I have a problem with my HA and Sonoff.
> 
> How your resolved?

This my error:

2019-11-04 10:49:21 ERROR (Thread-2) [custom_components.sonoff] websocket error: Connection is already closed.

Versions Sonoff 3.3.0
version HA: 0.99.3

**Comment by peterbuga** [comment]
@Tomelin i need more details for this, at some point if there's an error there should be more details about it (usually not just 1 line)

**Comment by rlcobos2** [comment]
Hi, I have the same error...

This occurs from the last update, i was some months without one single problem, but i need use the last commit for the 101 update of HA.

ERROR (Thread-2) [websocket] error from callback <bound method Sonoff.on_message of <custom_components.sonoff.Sonoff object at 0x71a11250>>: string indices must be integers

My code has the %s%s%s.

I will try to make a new ewelink account to share the devices between accounts.

any other suggestion??

**Comment by peterbuga** [comment]
@rlcobos2 update the component

**Comment by ijan1st** [comment]
ue Dec 17 2019 18:58:51 GMT+0800 (Malaysia Time)
Error loading custom_components.sonoff. Make sure all dependencies are installed
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/loader.py", line 346, in _load_file
    module = importlib.import_module(path)
  File "/usr/local/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/config/custom_components/sonoff/__init__.py", line 8, in <module>
    from homeassistant.util.async_ import run_coroutine_threadsafe
ImportError: cannot import name 'run_coroutine_threadsafe' from 'homeassistant.util.async_' (/usr/src/homeassistant/homeassistant/util/async_.py)


ive got this error after updating to hass-sonoff-ewelink 0.32
im running HA 0.103

**Comment by ijan1st** [comment]
websocket error: Connection is already closed.
8:02 PM custom_components/sonoff/__init__.py (ERROR) - message first occurred at 8:02 PM and shows up 2 times

anybody having this problem.. im using hass-sonoff-ewelink 0.3.3 and HA 0.103.. All my sonoff not working and showing this error.. any help plz.. tq

**Comment by ijan1st** [comment]
Sonoff debug
. --------------COPY-FROM-HERE--------------

2019-12-17 21:18:39.820 [S] {"entity_id": "1000359fa9", "new_state": "on", "outlet": 0}

2019-12-17 21:18:39.864 [W] {"error": 406, "sequence": "15765887198143337"}

2019-12-17 21:18:40.212 [W] {"actionName": "userOnline", "apikey": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", "error": 400, "reason": "Bad Request", "sequence": "15765887201624284"}

2019-12-17 21:18:43.276 [S] {"entity_id": "1000359fa9", "new_state": "on", "outlet": 1}

2019-12-17 21:18:43.322 [W] {"error": 406, "sequence": "15765887232747319"}

2019-12-17 21:18:43.668 [W] {"actionName": "userOnline", "apikey": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", "error": 400, "reason": "Bad Request", "sequence": "15765887236189022"}


this the sonoff debug..

**Comment by david78455** [comment]
Hi
Same problem nothing is working. I keep receive this error in the log
(Thread-2) [custom_components.sonoff] websocket error: Connection is already closed.
Please help
Regards
David,

**Comment by rlcobos2** [comment]
Hi, I have solved the issue updating the component, updating Hass.io , and, making a new account in ewelink sharing all the devices. Now i have two accounts in ewelink, one for the app, and other one, only for hassio.
